### PR TITLE
DDF-3920 Upgrade WSS4J and CXF

### PR DESCRIPTION
**Pull request by coyotesqrl**
#### What does this PR do?
Upgrades WSS4J and CXF dependencies.

#### Who is reviewing it? 
@tbatie @brjeter @roelens8 
(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)
#### Select relevant component teams: 
@codice/security 
#### Ask 2 committers to review/merge the PR and tag them here. If you don't know who to ask, you can request reviews in https://groups.google.com/forum/#!forum/ddf-developers .
(please choose ONLY two committers from below, delete the rest)
@rzwiefel
@stustison

#### How should this be tested? (List steps with links to updated documentation)
This should go through CI and thorough manual regression tests on all supported operating systems.

#### What are the relevant tickets?
[DDF-3920](https://codice.atlassian.net/browse/DDF-3920)
#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: features/pom.xml [hunk 1]
```diff
@@ -89,7 +89,7 @@
         <aries.util.version>1.1.3</aries.util.version>
         <!-- END enterprise properties -->
 
-        <!--START versions for CXF features in platform-app features
+        <!--START versions for CXF features
         Copied from apache-cxf-3.1.13-src/parent/pom.xml
```
Comment (emmberk) [inline features/pom.xml]: Should the version here be updated?

**Comment by coyotesqrl** [comment]
build now

(I want to get CI results early/often on this PR during the review process.)

File: pom.xml [hunk 2]
```diff
@@ -135,9 +140,11 @@
         <handlebars.version>2.0.0</handlebars.version>
         <antlr.version>4.3</antlr.version>
         <lux.woodstox.version>4.1.1</lux.woodstox.version>
-        <commons-codec.version>1.10</commons-codec.version>
+        <commons-codec.version>1.11</commons-codec.version>
         <spring.version>4.3.10.RELEASE</spring.version>
         <jetty.version>9.3.14.v20161028</jetty.version>
+        <javax.annotation-api.version>1.3</javax.annotation-api.version>
+        <opensaml.osgi.version>3.3.0_2</opensaml.osgi.version>
```
Comment (emmberk) [inline pom.xml]:
```xml
<javax.annotation-api.version>1.3</javax.annotation-api.version>
<opensaml.osgi.version>3.3.0_2</opensaml.osgi.version>
```
Do these 2 properties belong in the the `<!--START If these bundles are used elsewhere in DDF, the versions MUST exactly match the versions used by the CXF and Camel features.-->` section? If not, I don't think the version properties should be used in `features/pom.xml`:
```xml
<cxf.opensaml.osgi.version>1.3</cxf.opensaml.osgi.version>
<cxf.javax.servlet-api.version>3.3.0_2</cxf.javax.servlet-api.version>
```

File: features/camel/src/main/feature/feature.xml [hunk 3]
```diff
@@ -119,6 +119,7 @@
         <bundle>mvn:io.netty/netty-transport-native-epoll/${netty41-version}
         </bundle>
         <bundle>mvn:io.netty/netty-transport-native-kqueue/${netty.version}</bundle>
+        <bundle>mvn:io.netty/netty-transport-native-unix-common/${netty.version}</bundle>
```
Comment (coyotesqrl) [inline features/camel/src/main/feature/feature.xml]: This change here concerns me a small bit. The netty upgrade split out some package, clearly, but I'm unclear why the unix-common bundle really needed to be added, particularly as I was doing my development on OSX. This change, right here, is one of the most important ones to be tested across OSs.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-subjectconfirmation/pom.xml [hunk 4]
```diff
@@ -69,6 +69,10 @@
               platform-util,
               platform-util-unavailableurls
             </Embed-Dependency>
+            <Import-Package>
+              javax.annotation;version="[1.1,2)",
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-plugin/security-idp-plugin-subjectconfirmation/pom.xml]: I am not clear _why_ this needed to be added to this module directive.

File: platform/security/idp/security-idp-server/pom.xml [hunk 5]
```diff
@@ -247,6 +247,7 @@
                         <Import-Package>
                             !org.abego.treelayout.*,
                             com.sun.management;resolution:=optional,
+                            javax.annotation;version="[1.1,2)",
```
Comment (coyotesqrl) [inline platform/security/idp/security-idp-server/pom.xml]: I am not clear _why_ this needed to be added to this module directive.

File: platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java [hunk 6]
```diff
@@ -68,6 +72,20 @@
 
   private static final Logger LOGGER = LoggerFactory.getLogger(SimpleSign.class);
 
+  // Should be removed in favor of a lookup through third-party library when one is found.
+  private static final Map<String, String> RFC5698_DSSC_OID_TO_URI_MAP =
```
Comment (coyotesqrl) [inline platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java]: This change, and the rest in `SimpleSign`, are fallout from the library upgrades. They do improve DDF, by deferring to the _correct_ signature algorithm in more cases instead of falling back to SHA1withRSA/SHA1withDSA all the time.

File: features/pom.xml (hunk 1 above)
Comment (coyotesqrl) [reply features/pom.xml]: I should probably remove that piece of the comment. It's not strictly _true_.

File: pom.xml (hunk 2 above)
Comment (coyotesqrl) [reply pom.xml]: I'm not sure. I think I need to take a pass over the two big poms and clean up/clarify the properties.

File: features/pom.xml (hunk 1 above)
Comment (emmberk) [reply features/pom.xml]: I would argue to leave the comment and then document exactly what changes were made from the original `/parent/pom.xml` file. I would think that this would help with future upgrades.

File: features/cxf/src/main/feature/feature.xml [hunk 7]
```diff
@@ -541,10 +471,9 @@
         <feature version="${cxf.version}">cxf-jaxrs</feature>
         <feature version="${cxf.version}">cxf-databinding-jaxb</feature>
         <feature version="${cxf.version}">cxf-databinding-aegis</feature>
-        <feature version="${cxf.version}">cxf-databinding-xmlbeans</feature>
         <feature version="${cxf.version}">cxf-bindings-corba</feature>
         <feature version="${cxf.version}">cxf-bindings-coloc</feature>
-        <feature version="${cxf.version}">cxf-bindings-object</feature>
+        <!--<feature version="${cxf.version}">cxf-http-provider</feature>-->
         <feature version="${cxf.version}">cxf-http-jetty</feature>
```
Comment (coyotesqrl) [inline features/cxf/src/main/feature/feature.xml]: I should have added this line back in. It is not present in this version of CXF, as the `cxf-http-jetty` feature is presumed to start with `cxf-http-provider`.

File: platform/security/idp/security-idp-plugin/security-idp-plugin-subjectconfirmation/pom.xml (hunk 4 above)
Comment (tbatie) [reply platform/security/idp/security-idp-plugin/security-idp-plugin-subjectconfirmation/pom.xml]: I looked into this a bit more. It's coming from the dependency tree: ddf-security-common -> cxf-rt-frontend-jaxrs -> javax.annotations 1.3. I don't think we're starting up a bundle exporting javax.annotations 1.3 ¯\_(ツ)_/¯ Not sure if this will impact much. Was this failing at build or runtime? If at runtime,  we could start up the javax.annotations 1.3 bundle as well. Not sure if it's worth it if nothing is broken though.

Comment (coyotesqrl) [reply platform/security/idp/security-idp-plugin/security-idp-plugin-subjectconfirmation/pom.xml]: Runtime. It may not be necessary any longer, as I'm installing the annotations jar in the `lib/ext` directory.

File: features/cxf/src/main/feature/feature.xml [hunk 8]
```diff
@@ -46,43 +46,22 @@
...
-        <bundle start-level="20">
-            mvn:org.codehaus.woodstox/woodstox-core-asl/${cxf.woodstox.core.version}
-        </bundle>
-        <bundle start-level="20">
-            mvn:org.apache.servicemix.bundles/org.apache.servicemix.bundles.jaxb-impl/${cxf.jaxb.bundle.version}
-        </bundle>
-        <bundle start-level="20">
-            mvn:org.apache.servicemix.bundles/org.apache.servicemix.bundles.jaxb-xjc/${cxf.jaxb.bundle.version}
-        </bundle>
+        <bundle start-level="10" dependency="true">mvn:org.apache.servicemix.specs/org.apache.servicemix.specs.activation-api-1.1/${cxf.servicemix.specs.version}</bundle>
```
[the remark this answers is not in the document]
Comment (coyotesqrl) [reply features/cxf/src/main/feature/feature.xml]: You may be right, but the goal is to make the file as similar to the cxf feature file and they made this change.

**Comment by coyotesqrl** [comment]
Pushed minor changes to the cxf feature file - rewrote the comment block about changes to correct and clarify and added a handful of missing `dependency="true"` attributes to some bundles in feature defs.

**Comment by coyotesqrl** [comment]
build now

**Comment by coyotesqrl** [comment]
build now

(yarn error)

File: pom.xml (hunk 2 above)
Comment (emmberk) [reply pom.xml]: I'd still like to see this updated.

If these 2 version properties
```xml
<javax.annotation-api.version>1.3</javax.annotation-api.version>
<opensaml.osgi.version>3.3.0_2</opensaml.osgi.version>
```
are used in the cxf version properties
```xml
<cxf.javax.annotation-api.version>${javax.annotation-api.version}</cxf.javax.annotation-api.version>
<cxf.opensaml.osgi.version>${opensaml.osgi.version}</cxf.opensaml.osgi.version>
```
, they should be moved to the 
```xml
         <!--START If these bundles are used elsewhere in DDF, the versions MUST exactly match the versions used by the CXF and Camel features.-->
```
section. If they do not belong in that section, then the root pom version properties should not be used in `features/pom.xml`
 ```xml
<cxf.javax.annotation-api.version>1.3</cxf.javax.annotation-api.version>
<cxf.opensaml.osgi.version>3.3.0_2</cxf.opensaml.osgi.version>
```

File: pom.xml [hunk 9]
```diff
@@ -182,6 +180,8 @@
         <common-compress.version>1.14</common-compress.version>
         <owasp-html-sanitizer.version>20171016.1</owasp-html-sanitizer.version>
         <!--START If these bundles are used elsewhere in DDF, the versions MUST exactly match the versions used by the CXF and Camel features.-->
+        <javax.annotation-api.version>1.3</javax.annotation-api.version>
```
Comment (coyotesqrl) [inline pom.xml]: @emmberk properties moved per your request 👍

Comment (emmberk) [reply pom.xml]: Looks great. Thank you!

File: platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java [hunk 10]
```diff
@@ -188,17 +196,35 @@ public void signUriString(String queryParams, UriBuilder uriBuilder) throws Sign
...
+    String jceSigAlgo = certificate.getSigAlgName();
 
-    if (pubKeyAlgo.equalsIgnoreCase("DSA")) {
-      sigAlgo = SSOConstants.DSA_SHA1;
+    java.security.Signature signature;
+    try {
+      if ("DSA".equalsIgnoreCase(certificate.getPublicKey().getAlgorithm())) {
+        signature = java.security.Signature.getInstance(jceSigAlgo, "BC");
+      } else {
+        signature = java.security.Signature.getInstance(jceSigAlgo);
```
Comment (brjeter) [inline platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java]: ❓ Could we not just pass in BouncyCastle every time?

File: platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java [hunk 11]
```diff
@@ -188,17 +196,35 @@ public void signUriString(String queryParams, UriBuilder uriBuilder) throws Sign
...
 
-    LOGGER.debug("Using Signature algorithm {}", sigAlgo);
+  private String getSignatureAlgorithmURI(X509Certificate certificate) {
+    String sigAlgoUri = RFC5698_DSSC_OID_TO_URI_MAP.get(certificate.getSigAlgOID());
 
-    return sigAlgo;
+    if (sigAlgoUri == null) {
+      sigAlgoUri =
+          ("DSA".equalsIgnoreCase(certificate.getPublicKey().getAlgorithm()))
+              ? WSS4JConstants.SHA384
```
Comment (brjeter) [inline platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java]: ❓ Where did `SHA384` come from?

Comment (coyotesqrl) [reply platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java]: @stustison and I discussed and decided to use the more secure SHA384 as our fallback hashing algorithm for DSA.

File: features/cxf/src/main/feature/feature.xml [hunk 12]
```diff
@@ -106,24 +119,31 @@
...
+            mvn:commons-codec/commons-codec/${cxf.commons-codec.version}
         </bundle>
-        <bundle start-level="25">mvn:com.google.guava/guava/${cxf.guava.version}</bundle>
-        <bundle start-level="25">
+        <bundle start-level="25" dependency="true">
+            mvn:org.apache.santuario/xmlsec/${cxf.xmlsec.bundle.version}
+        </bundle>
+        <bundle start-level="25" dependency="true">mvn:com.google.guava/guava/${cxf.guava.version}
+        </bundle>
+        <bundle start-level="25" dependency='true'>mvn:io.dropwizard.metrics/metrics-core/3.2.6
```
[the remark this answers is not in the document]
Comment (coyotesqrl) [reply features/cxf/src/main/feature/feature.xml]: 👍

File: platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java [hunk 13]
```diff
@@ -188,17 +196,35 @@ public void signUriString(String queryParams, UriBuilder uriBuilder) throws Sign
     return signature;
   }
 
-  private String getSignatureAlgorithm(X509Certificate certificate) {
-    String sigAlgo = SSOConstants.RSA_SHA1;
-    String pubKeyAlgo = certificate.getPublicKey().getAlgorithm();
+  private java.security.Signature getSignature(X509Certificate certificate)
+      throws SignatureException {
+    String jceSigAlgo = certificate.getSigAlgName();
```
Comment (brjeter) [inline platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java]: ❗️ `getSigAlgName()` returns the signature algorithm that the CA used to sign this certificate. Need to call `certificate.getPublicKey().getAlgorithm()` here (which will just return the encrypting algorithm without the hashing algorithm i.e. "RSA") . And then we need to append the hashing algorithm that we want to use for the signing. `"SHA1with<encrypting-algo>`. NOTE: This should only be used for signing. For validating signatures, we can't use this method because we need to pull the hashing algorithm AND encrypting algorithm from the signature we're given.

File: platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java [hunk 14]
```diff
@@ -248,19 +274,13 @@ public boolean validateSignature(
           certificateFactory.generateCertificate(
               new ByteArrayInputStream(Base64.getMimeDecoder().decode(encodedPublicKey)));
 
-      String jceSigAlgo = "SHA1withRSA";
-      if ("DSA".equalsIgnoreCase(certificate.getPublicKey().getAlgorithm())) {
-        jceSigAlgo = "SHA1withDSA";
-      }
+      java.security.Signature sig = getSignature((X509Certificate) certificate);
 
-      java.security.Signature sig = java.security.Signature.getInstance(jceSigAlgo);
```
Comment (brjeter) [inline platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java]: ❗️ Technically this is out of scope of this PR since it was already broken, but we need to pull the SigAlg from the `queryParamsToValidate`. That has this SigAlg in it. So this code should read something like:
```
String jceSigAlgo = <parse SigAlg parameter from queryParamsToValidate>
java.security.Signature sig = java.security.Signature.getInstance(jceSigAlgo);
```

File: platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java (hunk 13 above)
Comment (coyotesqrl) [reply platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java]: I want to be sure I've got this right. @brjeter, @blen-desta, and @stustison, is this correct now to the best of your knowledge, where I've replaced the `getSignature` method with two methods, one for signing and one for verification?
```java
  private java.security.Signature getValidationSignature(X509Certificate certificate)
      throws SignatureException {
    String jceSigAlgo = certificate.getSigAlgName();

    java.security.Signature signature;
    try {
      if ("DSA".equalsIgnoreCase(certificate.getPublicKey().getAlgorithm())) {
        signature = java.security.Signature.getInstance(jceSigAlgo, "BC");
      } else {
        signature = java.security.Signature.getInstance(jceSigAlgo);
      }
    } catch (NoSuchAlgorithmException | NoSuchProviderException e) {
      throw new SignatureException(e);
    }
    return signature;
  }

  private java.security.Signature getSigningSignature(X509Certificate certificate)
      throws SignatureException {
    java.security.Signature signature;
    try {
      if ("DSA".equalsIgnoreCase(certificate.getPublicKey().getAlgorithm())) {
        signature = java.security.Signature.getInstance("SHA384withDSA", "BC");
      } else {
        signature = java.security.Signature.getInstance("SHA384withRSA");
      }
    } catch (NoSuchAlgorithmException | NoSuchProviderException e) {
      throw new SignatureException(e);
    }
    return signature;
  }
```

Comment (brjeter) [reply platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java]: The `getValidationSignature` method won't be necessary. When validating a signature we need to be given a signature algorithm rather than pulling it off the cert. For Redirect, it's given in the queryParams (see my second comment). For POST, it's already in the OpenSAML `Signature` object that's passed in.

Comment (stustison) [reply platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java]: What @brjeter said and also we are probably fine with just using the SHA256 signatures

**Comment by coyotesqrl** [comment]
build now

File: features/cxf/src/main/feature/feature.xml [hunk 15]
```diff
@@ -233,7 +264,7 @@
         <!-- Can be used instead cxf-http for a smaller footprint -->
         <feature version="${cxf.version}">cxf-core</feature>
         <bundle start-level="40">
-            mvn:${cxf.servlet-api.group}/${cxf.servlet-api.artifact}/${cxf.servlet-api.version}
+            mvn:javax.servlet/javax.servlet-api/${cxf.javax.servlet-api.version}
```
Comment (ahoffer) [inline features/cxf/src/main/feature/feature.xml]: How is it that we got to move off the CXF API and onto the javax version of the API?

File: features/cxf/src/main/feature/feature.xml [hunk 16]
```diff
@@ -365,66 +434,87 @@
...
+        <bundle start-level="35">mvn:io.zipkin.brave/brave-instrumentation-http/4.6.0</bundle>
+        <bundle start-level="40">mvn:org.apache.cxf/cxf-integration-tracing-brave/${cxf.version}
+        </bundle>
+    </feature>
+    <feature name="cxf-tracing-opentracing" version="${cxf.version}">
+        <feature version="${cxf.version}">cxf-core</feature>
+        <bundle start-level="35">wrap:mvn:io.opentracing/opentracing-api/0.30.0</bundle>
+        <bundle start-level="35">wrap:mvn:io.opentracing/opentracing-util/0.30.0</bundle>
+        <bundle start-level="40">
+            mvn:org.apache.cxf/cxf-integration-tracing-opentracing/${cxf.version}
```
Comment (ahoffer) [inline features/cxf/src/main/feature/feature.xml]: Is the tracing/brave new? Will it improve debugging?

**Comment by coyotesqrl** [comment]
Build now

Nexus issue

File: features/cxf/src/main/feature/feature.xml [hunk 17]
```diff
@@ -365,66 +434,87 @@
...
         </bundle>
     </feature>
     <feature name="cxf-rs-description-swagger2" version="${cxf.version}">
         <feature version="${cxf.version}">cxf-jaxrs</feature>
         <feature version="${cxf.version}">cxf-jackson</feature>
+        <bundle start-level="35">
+            mvn:org.apache.cxf/cxf-rt-rs-service-description-swagger-ui/${cxf.version}
+        </bundle>
```
Comment (ahoffer) [inline features/cxf/src/main/feature/feature.xml]: Swagger integration?  Will it replace what we see when we hit /services?

File: features/cxf/src/main/feature/feature.xml (hunk 16 above)
Comment (coyotesqrl) [reply features/cxf/src/main/feature/feature.xml]: Different kind of tracing. We are not currently using it; I could not say for sure if other CXF libraries we use are but suspect they do.

File: features/cxf/src/main/feature/feature.xml (hunk 17 above)
Comment (coyotesqrl) [reply features/cxf/src/main/feature/feature.xml]: Again, most likely just in use by other CXF libraries.

**Comment by coyotesqrl** [comment]
build now

Upgraded to cxf 3.2.5 and fixed issues with `SimpleSign`.

**Comment by peterhuffer** [comment]
looks like last failure was problem with CI environment
build now

**Comment by Lambeaux** [comment]
Build now

**Comment by coyotesqrl** [comment]
build now

- rebased and resolved with geotools upgrade

**Comment by coyotesqrl** [comment]
build now

- was unable to reproduce locally and have seen a few other examples of potentially flaky environmental behavior on past PRs

**Comment by brjeter** [comment]
@blen-desta and I have hero'd this to the best of our ability.

**Comment by coyotesqrl** [comment]
Sidestepping static analysis and CI build as manual heroing process has verified.
